exporter.go

```diff
@@ -110,6 +113,43 @@ func createClientWithRetries(getClient func() (interface{}, error), retries uint
 	return nil, err
 }
 
+func parseUnixSocketAddress(address string) (string, string, error) {
+	addressParts := strings.Split(address, ":")
+	addressPartsLength := len(addressParts)
+
+	if addressPartsLength > 3 || addressPartsLength < 1 {
+		return "", "", fmt.Errorf("address for unix domain socket has wrong format")
+	}
+
+	unixSocketPath := addressParts[1]
+	requestPath := ""
+	if addressPartsLength == 3 {
+		requestPath = addressParts[2]
+	}
+	return unixSocketPath, requestPath, nil
+}
+
+func getListener(listenAddress string) (net.Listener, error) {
+	var listener net.Listener
+	var err error
+
+	if strings.HasPrefix(listenAddress, "unix:") {
+		path, _, err := parseUnixSocketAddress(listenAddress)
+		if err != nil {
+			return listener, fmt.Errorf("parsing unix domain socket listen address %s failed: %v", listenAddress, err)
+		}
+		listener, err = net.ListenUnix("unix", &net.UnixAddr{Name: path, Net: "unix"})
```

Comment:
the `err` in this line is a variable defined in line 137, not the line 134 (`var err error`). As a result, if `net.ListenUnix("unix", &net.UnixAddr{Name: path, Net: "unix"})` returns an error, it will not get caught in line 146:
```
	if err != nil {
		return listener, err
	}
```

That's why we got a cryptic message when the exporter tried to bind to the existing unix domain socket:

```
$ ./nginx-prometheus-exporter -nginx.scrape-uri=http://demo.nginx.com/stub_status -web.listen-address=unix:./socket
2019/11/07 21:39:26 Starting NGINX Prometheus Exporter Version=0.4.2 GitCommit=56b8d95
2019/11/07 21:39:26 NGINX Prometheus Exporter has successfully started
2019/11/07 21:39:26 Listening on unix:./socket
2019/11/07 21:39:26 invalid argument
```

If you fix the bug, you will get:
```
2019/11/08 17:11:43 Could not create listener: listen unix socket-exporter: bind: address already in use
```

Reply:
Yes, I made a mistake here. It's correct now 😉.